google/cloud/storage/internal/service_account_credentials.h

```diff
@@ -0,0 +1,183 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_SERVICE_ACCOUNT_CREDENTIALS_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_SERVICE_ACCOUNT_CREDENTIALS_H_
+
+#include "google/cloud/storage/credentials.h"
+#include "google/cloud/storage/internal/credential_constants.h"
+#include "google/cloud/storage/internal/curl_request.h"
+#include "google/cloud/storage/internal/nljson.h"
+#include "google/cloud/storage/internal/openssl_util.h"
+#include "google/cloud/storage/internal/time_fetcher.h"
+#include <chrono>
+#include <condition_variable>
+#include <ctime>
+#include <iostream>
+#include <mutex>
+#include <sstream>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+/**
+ * A C++ wrapper for Google's Service Account Credentials.
+ *
+ * Takes a JSON object representing the contents of a service account keyfile,
+ * and uses Google's OAuth2 service to obtain an access token.
+ *
+ * @par Warning
+ * The current implementation is a placeholder to unblock development of the
+ * Google Cloud Storage client libraries. There is substantial work needed
+ * before this class is complete, in fact, we do not even have a complete set of
+ * requirements for it.
+ *
+ * @see
+ *   https://developers.google.com/identity/protocols/OAuth2ServiceAccount
+ *   https://tools.ietf.org/html/rfc7523
+ *
+ * @tparam HttpRequestType a dependency injection point to make HTTP requests.
+ */
+template <typename HttpRequestType = CurlRequest,
+          typename TimeFetcherType = TimeFetcher>
+class ServiceAccountCredentials : public storage::Credentials {
+ public:
+  explicit ServiceAccountCredentials(std::string const& content)
+      : ServiceAccountCredentials(content, GoogleOAuthRefreshEndpoint()) {}
+
+  explicit ServiceAccountCredentials(std::string const& content,
+                                     std::string oauth_server)
+      : requestor_(std::move(oauth_server)),
+        expiration_time_(),
+        time_fetcher_() {
+    auto credentials = nl::json::parse(content);
+    // This is the value of grant_type for:
+    // - JSON-formatted service account keyfiles downloaded from Cloud Console
+    nl::json assertion_header = {
+        {"alg", "RS256"},
+        {"kid", credentials["private_key_id"].get_ref<std::string const&>()},
+        {"typ", "JWT"}};
+
+    std::ostringstream scope_oss;
+    scope_oss << GoogleOAuthScopeCloudPlatform() << " "
```

Comment:
Shouldn't you only need the google cloud platform scope here?

Reply:
I was just going off of what the oauth2client code (called from gsutil) did.  I could probably just use the cloud platform scope here, though - you're right. I think changing the scopes would result in altering the content of the JWT refresh request, which would mean recomputing the expected values in service_account_credentials_test.cc... As I'd like to get this PR pushed before I go on vacation (after tomorrow) are you cool with me creating an issue to fix this (and the tests) in a couple weeks?